File: GoEngineering/14_concurrent_cache/explanation/14_concurrent_cache_explanation.go
/*
Problem 14 — Concurrent Cache (sync.Map vs mutex, single-flight, TTL eviction)

# WHAT WE'RE BUILDING

A generic, concurrency-safe, in-process cache with:
  - TTL-based expiry (entries become invisible/collectible after their
    time-to-live elapses, with a background sweeper reclaiming memory
    instead of relying purely on lazy get-time checks),
  - single-flight loading (concurrent Gets for the same missing key
    collapse into exactly one call to the loader function, not N),
  - a choice of two internal storage strategies (mutex-protected map vs
    sync.Map) selectable at construction, so the trade-off is something you
    can benchmark, not just read about.

This is the cache shape that sits in front of an expensive operation inside
a single Go process — a DB row, a computed aggregate, a downstream API
response — where the goal is "never do the expensive work twice
concurrently for the same key, and never serve stale data past its TTL."

# WHY THIS MATTERS IN REAL SYSTEMS

Three failure modes show up constantly in hand-rolled caches:

 1. **Cache stampede / thundering herd.** A popular key expires (or is
    never populated) and 500 concurrent requests all miss at once, all call
    the expensive loader at once, and the backing store that the cache was
    protecting gets hit with 500x its normal load simultaneously. This is
    one of the most common real production incidents caused by caching
    code. Single-flight collapses those 500 calls into 1: the other 499
    goroutines block briefly and then receive the same result the 1
    in-flight call produced.
 2. **Unbounded growth from lazy-only expiry.** If entries are only checked
    "is this expired?" when read, a key that stops being read (e.g. a user
    who never comes back) stays in memory forever — the map only shrinks on
    access, never on time passing. A background sweeper (or an eviction
    check woven into every operation, if you choose to avoid a goroutine)
    is what actually bounds memory over the life of a long-running process.
 3. **Choosing sync.Map by default.** sync.Map is optimized for a specific
    access pattern (keys mostly written once and read many times by many
    goroutines, or disjoint key sets per goroutine) and is a Pessimization
    for a general-purpose read-write-heavy cache — the stdlib docs say this
    explicitly. A mutex-protected map[K]V is faster for most cache
    workloads. This exercise makes you implement both and benchmark them so
    the trade-off is measured, not assumed.

# CONCEPTS COVERED

  - Generic cache: `Cache[K comparable, V any]`
  - sync.Map vs sync.Mutex+map: two Store implementations behind one
    interface, benchmarked against each other
  - Hand-rolled single-flight (do NOT import golang.org/x/sync/singleflight
    — implement the mechanism yourself so you understand what it does: a
    map of in-flight calls keyed by K, each guarded so concurrent callers
    for the same key wait on the SAME loader invocation's result instead of
    starting their own)
  - TTL eviction: per-entry expiry timestamp + a background sweeper
    goroutine with clean shutdown (no goroutine leak)
  - Context-aware loading: the loader function accepts a context so a
    caller's cancellation/timeout propagates into (or at least bounds) the
    single-flight-shared load
  - Race safety verified with `go test -race`

# SPEC

	type Loader[K comparable, V any] func(ctx context.Context, key K) (V, error)

	type Cache[K comparable, V any] struct { ... }

	type Store string // "mutex" or "syncmap" — selects internal storage strategy

	type Options struct {
	    TTL             time.Duration // required, > 0
	    SweepInterval   time.Duration // how often the background sweeper runs; > 0
	    Store           Store         // defaults to "mutex" if empty
	}

	func New[K comparable, V any](opts Options) *Cache[K, V]
	    Constructs a ready-to-use cache and starts its background sweeper
	    goroutine.

	func (c *Cache[K, V]) Get(ctx context.Context, key K, load Loader[K, V]) (V, error)
	    Returns the cached value for key if present and not expired.
	    Otherwise calls load exactly once per set of concurrently-missing
	    callers for that key (single-flight), stores the result (only on
	    success — a loader error is never cached) with a fresh TTL, and
	    returns it to every waiter.

	func (c *Cache[K, V]) Set(key K, value V)
	    Unconditionally stores value for key with a fresh TTL, bypassing
	    the loader path entirely (for pre-warming or explicit writes).

	func (c *Cache[K, V]) Delete(key K)
	    Removes key immediately, regardless of TTL.

	func (c *Cache[K, V]) Len() int
	    Number of currently live (non-expired) entries. For the sync.Map
	    backend this necessarily means a full scan — document that cost.

	func (c *Cache[K, V]) Close()
	    Stops the background sweeper goroutine. Idempotent. After Close,
	    Get/Set/Delete remain safe to call (they just won't benefit from
	    background sweeping — lazy expiry-on-read still applies) — Close
	    is a resource-cleanup operation, not a hard shutdown gate.

ACCEPTANCE CRITERIA

  - N concurrent Get calls for the same missing key result in exactly ONE
    call to the loader function (verified in tests with an atomic call
    counter).
  - An expired entry is never returned by Get: a lazy check-on-read backstop
    catches anything the background sweeper hasn't gotten to yet.
  - The background sweeper does not leak: Close() stops it, verified with a
    done-channel or similar deterministic synchronization (not a sleep).
  - Race-free under `go test -race` with concurrent Get/Set/Delete/Close
    from many goroutines against the same cache, for both Store variants.
  - A loader error is propagated to all waiters for that single-flight call
    and is NOT cached — the next Get retries the loader.
*/
package cache

import (
	"context"
	"time"
)

// Loader loads the value for key on a cache miss.
type Loader[K comparable, V any] func(ctx context.Context, key K) (V, error)

// Store selects the internal storage strategy.
type Store string

const (
	// StoreMutex backs the cache with a plain map[K]entry[V] guarded by a
	// sync.Mutex. TODO: this is the default when Options.Store is empty.
	StoreMutex Store = "mutex"
	// StoreSyncMap backs the cache with sync.Map.
	StoreSyncMap Store = "syncmap"
)

// Options configures a new Cache.
type Options struct {
	// TTL is how long an entry stays valid after being set. Required, > 0.
	TTL time.Duration
	// SweepInterval is how often the background sweeper scans for and
	// removes expired entries. Required, > 0.
	SweepInterval time.Duration
	// Store selects StoreMutex or StoreSyncMap. Zero value defaults to
	// StoreMutex.
	Store Store
}

// Cache is a generic, concurrency-safe, TTL-expiring cache with
// single-flight loading. The zero value is not usable; construct with New.
//
// TODO: fields you'll likely need:
//   - opts Options
//   - a storage abstraction (interface or two concrete implementations
//     selected at construction time) holding entry[V] per key
//   - inflight map[K]*call[V] + a mutex guarding it, for single-flight
//   - a stop channel + sync.WaitGroup (or similar) for Close() to
//     shut down the sweeper goroutine deterministically
type Cache[K comparable, V any] struct {
	// TODO: fields
}

// entry is one stored value plus its expiry deadline.
// TODO: fields: value V, expiresAt time.Time.
type entry[V any] struct {
	// TODO: fields
}

// call is one in-flight single-flight load for a given key: the first
// goroutine to miss on a key creates one of these, does the actual load,
// and every other concurrently-missing goroutine for the SAME key waits on
// it via done instead of calling load itself.
// TODO: fields: done chan struct{} (closed when the load finishes), value
// V, err error — the classic "future" shape.
type call[V any] struct {
	// TODO: fields
}

// New constructs a ready-to-use Cache and starts its background sweeper
// goroutine. Panics if opts.TTL <= 0 or opts.SweepInterval <= 0.
//
// TODO: implement.
func New[K comparable, V any](opts Options) *Cache[K, V] {
	panic("TODO: implement New")
}

// Get returns the cached value for key if present and unexpired. On a
// miss, it calls load — collapsing concurrent misses for the same key into
// exactly one loader call (single-flight) — stores the result on success
// with a fresh TTL, and returns it to every waiter. A loader error is
// returned to every waiter and is never cached.
//
// TODO: implement:
//  1. Fast path: look up key; if present and not expired, return it.
//  2. Slow path: join or create the in-flight call for key under the
//     inflight-map's own lock (NOT the storage lock — see hints).
//     - If you created it: run load, store on success, close done,
//     release the inflight-map entry.
//     - If you joined an existing one: wait on done (respecting ctx
//     cancellation via select), then read the shared result.
func (c *Cache[K, V]) Get(ctx context.Context, key K, load Loader[K, V]) (V, error) {
	panic("TODO: implement Cache.Get")
}

// Set unconditionally stores value for key with a fresh TTL.
// TODO: implement.
func (c *Cache[K, V]) Set(key K, value V) {
	panic("TODO: implement Cache.Set")
}

// Delete removes key immediately, regardless of TTL.
// TODO: implement.
func (c *Cache[K, V]) Delete(key K) {
	panic("TODO: implement Cache.Delete")
}

// Len reports the number of currently live (non-expired) entries.
// TODO: implement. For the sync.Map backend this requires a full Range
// scan — document that cost at the call site or in the method doc.
func (c *Cache[K, V]) Len() int {
	panic("TODO: implement Cache.Len")
}

// Close stops the background sweeper goroutine. Idempotent; safe to call
// more than once. Get/Set/Delete remain safe to call afterward.
// TODO: implement using a sync.Once plus closing a stop channel, and wait
// (via sync.WaitGroup or a done channel) for the sweeper goroutine to
// actually exit before returning — Close should be synchronous, not
// "requested and hopefully done eventually."
func (c *Cache[K, V]) Close() {
	panic("TODO: implement Cache.Close")
}

/*
HINTS

  - Single-flight needs its OWN lock, separate from whatever protects the
    storage map. If you reuse the storage lock for the inflight bookkeeping
    too, you'll either deadlock (holding the storage lock while waiting on
    a load that itself wants the storage lock to Set the result) or
    serialize unrelated keys behind one key's slow load.
  - The "join or create" step for single-flight must be atomic: two
    goroutines checking "is there an inflight call for key?" and both
    finding none, then both creating one, defeats the entire point. Do the
    check-and-maybe-create under the inflight mutex as one critical
    section.
  - For sync.Map: LoadOrStore is your friend for atomic check-and-set of
    the *call[V] entry itself — but note sync.Map's LoadOrStore always
    evaluates its second argument eagerly (you're constructing a fresh
    *call[V] every time even when it won't be used) which is a real,
    measurable cost the mutex+map version doesn't pay; worth noting when
    you benchmark.
  - Context cancellation while WAITING on someone else's in-flight call
    should return ctx.Err() to the waiter that gave up, WITHOUT canceling
    the load for the other waiters still waiting on it — the load isn't
    "yours" to cancel just because you stopped waiting.
  - The background sweeper should hold its lock only long enough to scan
    and delete expired keys, never for the whole SweepInterval — a
    sweep is a bounded, fast operation; don't let it block Get/Set for
    unrelated keys any longer than necessary.

COMMON PITFALLS

  - Caching a loader error — the acceptance criteria explicitly forbid
    this: a transient failure (backing store hiccup) must not poison the
    cache for the full TTL; the next Get should retry.
  - Forgetting the lazy expiry-on-read backstop and relying only on the
    sweeper — with a SweepInterval longer than TTL (a legitimate config,
    e.g. sweep every minute, TTL 10s), Get would return stale data for up
    to a minute without the read-time check.
  - Closing the `done` channel on a *call[V] more than once (if load
    somehow runs twice for the same call — shouldn't happen with correct
    single-flight, but guard it) — panics.
  - Leaking the sweeper goroutine because Close() only sets a flag/closes a
    channel without waiting for the goroutine to observe it and exit —
    tests that check goroutine counts before/after Close will catch this,
    but only if you write that test.
  - Using time.Now() directly scattered through the code instead of
    capturing it once per operation — makes expiry-boundary tests flaky
    (two time.Now() calls microseconds apart can straddle the TTL boundary
    differently than intended).

STRETCH GOALS

  - Add an eviction policy on top of TTL (LRU or LFU with a max entry
    count) so the cache is bounded by size, not just by time.
  - Add per-key TTL overrides (Set taking an optional ttl parameter distinct
    from the cache-wide default).
  - Write a benchmark (`go test -bench`) comparing StoreMutex vs
    StoreSyncMap under: (a) read-heavy/write-rare workload, (b) balanced
    read/write, (c) write-heavy workload — confirm or refute the stdlib's
    documented sync.Map sweet spot on this machine.
  - Implement a real golang.org/x/sync/singleflight-style Do/Forget API
    surface (separate group keyed independently of the cache) and compare
    it against your hand-rolled version's behavior under cancellation.
*/

File: PyEngineering/14_concurrent_cache/14_concurrent_cache_explanation.py
"""
14 · Concurrent cache
=====================

WHAT WE'RE BUILDING
--------------------
A thread-safe, in-process cache with TTL eviction and single-flight
de-duplication: `get_or_compute(key, compute)` returns a cached value if
fresh, else computes it exactly once even when many threads race to
compute the same missing key at the same moment. This is the shape behind
`functools.lru_cache` in production services that actually need
expiration and cross-thread safety — a config cache, a per-request
memoized DB lookup, a computed-feature cache in front of a slow model.

WHY THREADS, NOT ASYNCIO, FOR THIS ONE
-----------------------------------------
Every other concurrency problem in this curriculum (12, 13, 15, 16) is
`asyncio`-native, because the concurrency is I/O-bound and single-threaded
cooperative scheduling is the right tool. A cache is different: it is
routinely called from **synchronous code paths** that have no event loop
at all — a Django/Flask request handler, a Celery task, a plain function
deep in a call stack that an `asyncio` service calls via
`run_in_executor`. Locking primitives don't cross the async/sync boundary
cleanly (`asyncio.Lock` only works correctly when awaited from inside the
same running event loop — a `threading.Lock`-protected structure is
usable from any thread, sync or async-via-executor, with no ownership
constraints). So: `threading.Lock` here, deliberately, because the real
in-production version of this exact component overwhelmingly is a plain
thread-safe class, not an asyncio primitive.

WHY THIS MATTERS IN REAL SYSTEMS
---------------------------------
- **The thundering herd problem.** Without single-flight, N threads that
  all miss the cache for the same key at the same moment each independently
  call the (possibly expensive/rate-limited/billed-per-call) `compute`
  function — N calls instead of 1. This is a classic real incident:
  a cache entry expires under load, and the recompute stampede takes down
  the backing database/API it was protecting. Single-flight collapses the
  N racing callers onto exactly one in-flight computation; every caller
  (including the N-1 that didn't win the race) gets the same result.
- **TTL eviction has a lazy-vs-eager trade-off.** Pure lazy expiry (check
  the timestamp on every `get`, evict then) never wastes CPU cleaning
  entries nobody's asking for, but a cache that's mostly cold-written and
  rarely re-read leaks expired entries in memory forever. An optional
  background sweep thread trades a small constant CPU/wakeup cost for a
  bounded memory ceiling — the right choice depends on the access pattern,
  and real systems often ship both (lazy expiry for correctness, periodic
  sweep for memory bounds), which is what this problem does.
- **Lock scope is the whole game.** Holding the cache's lock while calling
  `compute()` (which the caller supplies and may be slow/network-bound)
  would serialize *every* cache access — hits included — behind the
  slowest in-flight miss. Correct single-flight needs a lock per *key*
  during computation, held only long enough to register/await the
  in-flight computation, not the cache's single global lock held for the
  duration of the call.

CONCEPTS COVERED
-----------------
- `threading.Lock` for structure integrity vs. a hand-rolled single-flight
  mechanism (a per-key `threading.Event`-gated slot, not a library)
- Lazy TTL expiry (checked on read) + an optional daemon background sweep
  thread for proactive cleanup
- Why holding a lock across a slow/user-supplied call is a bug, and how to
  avoid it while still guaranteeing exactly-once computation

THE SPEC
--------
Build `ConcurrentCache[K, V]` (PEP 695 generic class):

    class ConcurrentCache[K, V]:
        def __init__(
            self,
            *,
            ttl_seconds: float | None = None,
            sweep_interval_seconds: float | None = None,
        ) -> None: ...
        def get_or_compute(self, key: K, compute: Callable[[], V]) -> V: ...
        def get(self, key: K) -> V | None: ...
        def invalidate(self, key: K) -> None: ...
        def __len__(self) -> int: ...
        def close(self) -> None: ...
            # stops the background sweep thread if one was started

Behavior:
    - `get_or_compute(key, compute)`: return the cached value if present
      and unexpired. Otherwise, ensure `compute()` runs *exactly once*
      across however many threads concurrently miss on the same key —
      every caller (the one that ran `compute` and every other one that
      raced it) receives the same computed value. If `compute()` raises,
      every waiting caller sees that same exception re-raised (the failure
      is not cached — the next call retries).
    - `ttl_seconds=None` means entries never expire.
    - `get(key)` is a read-only lookup: returns `None` on miss/expiry,
      never triggers computation, never blocks on an in-flight computation
      for that key (it should not wait for another thread's `compute`).
    - `invalidate(key)` removes an entry (no-op if absent).
    - If `sweep_interval_seconds` is set, a daemon background thread
      periodically scans and evicts expired entries (in addition to the
      lazy check every `get`/`get_or_compute` already does). `close()`
      stops it deterministically (joinable, no lingering thread after
      `close()` returns).

ACCEPTANCE CRITERIA
--------------------
1. Sequential correctness: cached values round-trip; TTL expiry causes a
   miss (and recompute) after the TTL elapses, verified with a fake/
   monkeypatched clock (not a real `time.sleep`, to keep tests fast and
   non-flaky).
2. **Single-flight, proven directly**: spin up many threads all calling
   `get_or_compute` on the same missing key simultaneously (synchronized
   via a `threading.Barrier`); assert the compute function's call count is
   exactly 1 and every thread receives the identical result.
3. A `compute()` exception is propagated to every racing caller and is
   NOT cached — the next `get_or_compute` call retries computation.
4. `get()` never blocks on another thread's in-flight `get_or_compute` for
   the same key.
5. With `sweep_interval_seconds` set, expired entries are removed by the
   background thread even without any `get` calls touching them; `close()`
   leaves no running non-daemon thread behind.
6. mypy-clean, black-formatted, ruff-clean.
"""

from __future__ import annotations

import threading
from collections.abc import Callable
from typing import Generic, TypeVar

K = TypeVar("K")
V = TypeVar("V")


class ConcurrentCache(Generic[K, V]):
    """Thread-safe cache: TTL eviction (lazy + optional sweep) + single-flight."""

    def __init__(
        self,
        *,
        ttl_seconds: float | None = None,
        sweep_interval_seconds: float | None = None,
    ) -> None:
        # TODO: self._ttl = ttl_seconds
        # TODO: self._lock = threading.Lock()  # protects the entries dict
        #       and the in-flight-computation registry, briefly, per call
        # TODO: self._entries: dict[K, tuple[V, float]] = {}  # value, expires_at
        # TODO: self._inflight: dict[K, "_SingleFlight[V]"] = {}
        # TODO: if sweep_interval_seconds is set, start a daemon
        #       threading.Thread running a sweep loop; store a
        #       threading.Event to signal it to stop, and the Thread
        #       object itself so close() can join() it.
        raise NotImplementedError("TODO: implement ConcurrentCache.__init__")

    def get(self, key: K) -> V | None:
        """Read-only lookup. Never computes, never blocks on another
        thread's in-flight computation for this key."""
        # TODO: under self._lock, look up key; if present and unexpired,
        # return the value; if present and expired, delete it and fall
        # through to return None.
        raise NotImplementedError("TODO: implement ConcurrentCache.get")

    def get_or_compute(self, key: K, compute: Callable[[], V]) -> V:
        """Return the cached value, or compute it exactly once under
        concurrent callers racing on the same missing key."""
        # TODO: this is the core of the problem. Sketch:
        # 1. Fast path: under the lock, if key present+unexpired, return it.
        # 2. Still under the lock: check self._inflight for key. If another
        #    thread is already computing this key, grab its _SingleFlight
        #    handle and release the lock, then WAIT on it (outside the
        #    lock!) and return/raise based on its outcome.
        # 3. Otherwise, still under the lock, create a fresh _SingleFlight
        #    and register it in self._inflight[key] — this "reserves" the
        #    key for this thread before releasing the lock, which is what
        #    makes step 2 correct for every thread that arrives after this
        #    point but before the computation finishes.
        # 4. Release the lock, then actually call compute() OUTSIDE the
        #    lock (never hold the cache lock across caller-supplied code).
        # 5. On success: store the value+expiry under the lock, remove the
        #    inflight entry, and wake every waiter with the result.
        #    On exception: remove the inflight entry (do NOT cache the
        #    failure) and wake every waiter with the exception re-raised.
        raise NotImplementedError("TODO: implement ConcurrentCache.get_or_compute")

    def invalidate(self, key: K) -> None:
        # TODO: under the lock, self._entries.pop(key, None).
        raise NotImplementedError("TODO: implement ConcurrentCache.invalidate")

    def __len__(self) -> int:
        # TODO: under the lock, return len(self._entries).
        raise NotImplementedError("TODO: implement ConcurrentCache.__len__")

    def close(self) -> None:
        """Stop the background sweep thread, if any, deterministically."""
        # TODO: if a sweep thread was started, set its stop-Event and
        # thread.join() it. Idempotent — safe to call twice.
        raise NotImplementedError("TODO: implement ConcurrentCache.close")


# ---------------------------------------------------------------------------
# HINTS
# -----
# - The single-flight handle needs exactly one thing beyond storage for
#   the result/exception: a way for N-1 threads to block until the 1
#   computing thread is done. `threading.Event` is the natural fit —
#   `.wait()` for waiters, `.set()` by the computing thread when finished.
# - The trickiest correctness property: the "reservation" (registering the
#   in-flight handle) must happen BEFORE the lock is released, and the
#   actual `compute()` call must happen AFTER the lock is released. Get
#   this ordering wrong in either direction and you either serialize all
#   cache access behind slow computes, or open a race window where two
#   threads both think they're the one computing.
# - For the fake-clock TTL test, inject a callable (`now: Callable[[],
#   float] = time.monotonic`) as a constructor parameter so tests can pass
#   a controllable fake instead of sleeping in real time.
#
# COMMON PITFALLS
# ---------------
# - Holding `self._lock` while calling `compute()` — this serializes the
#   entire cache (even unrelated keys, even pure hits) behind one slow
#   miss. The lock must be released before calling into caller code.
# - Caching a `compute()` exception as if it were a valid value — a
#   transient failure (a downstream timeout, say) must not permanently
#   poison the cache entry; the next call should retry.
# - A single-flight implementation that only de-duplicates *within* one
#   call to `get_or_compute` (e.g. a naive check-then-act with no
#   registration step) rather than truly across concurrently racing
#   threads — this is the bug the barrier-synchronized test is designed to
#   catch, and it's easy to write a version that passes a casual manual
#   test but fails under a real race.
# - Forgetting the sweep thread must be a daemon thread (or explicitly
#   joined in `close()`) — a non-daemon thread that's never stopped keeps
#   the whole process alive past `main()` returning, which is exactly the
#   kind of "leaked resource" this curriculum's problems keep testing for.
#
# STRETCH GOALS
# --------------
# - Add a maximum size with LRU eviction (an `OrderedDict` with
#   `move_to_end` on access, evicting from the front when over capacity)
#   and measure hit rate under a Zipfian access pattern.
# - Add per-key locks (a `dict[K, threading.Lock]`) instead of one global
#   structural lock, and measure whether it actually helps throughput at
#   realistic key cardinalities (Python's GIL means pure-CPU compute won't
#   benefit; I/O-bound compute — the realistic case — might).
# - Add `async def aget_or_compute` as a thin asyncio-facing wrapper via
#   `loop.run_in_executor`, and explain in comments why this is usually
#   the right integration point rather than rewriting the cache itself in
#   `asyncio` primitives.
